src/System.Private.CoreLib/shared/System/BitOps.cs

```diff
@@ -5,34 +5,747 @@
+        /// Similar in behavior to the x86 instruction LZCNT.
+        /// </summary>
+        /// <param name="value">The value.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static uint LeadingZeroCount(ulong value)
+        {
+            if (Lzcnt.X64.IsSupported)
+            {
+                return (uint)Lzcnt.X64.LeadingZeroCount(value);
+            }
+
+            // Use the 32-bit function twice.
+
+            uint hi = (uint)(value >> 32);
+            uint lo = (uint)value;
+
+            if (Lzcnt.IsSupported)
+            {
+                hi = Lzcnt.LeadingZeroCount(hi);
+                lo = Lzcnt.LeadingZeroCount(lo);
+            }
+            else
+            {
+                hi = LeadingZeroCount(hi);
+                lo = LeadingZeroCount(lo);
+            }
+
+            // Keep lo iff hi==32
+            uint m = hi & ~32u; // Zero 5th bit of hi
+            lo *= IsZero(m); // lo *= (m == 0 ? 1 : 0)
+
+            return hi + lo;
+        }
+
+        /* Legacy implementations
+        DONE https://raw.githubusercontent.com/dotnet/corefx/151a6065fa8184feb1ac4a55c89752342ab7c3bb/src/Common/src/CoreLib/System/Decimal.DecCalc.cs
+        DONE https://raw.githubusercontent.com/dotnet/corefx/62c70143cfbb08bbf03b5b8aad60c2add84a0d9e/src/Common/src/CoreLib/System/Number.BigInteger.cs
+        */
+
+        #endregion
+
+        #region TrailingZeroCount
+
+        // PR already merged: https://github.com/dotnet/coreclr/pull/22118
+
+        /// <summary>
+        /// Count the number of trailing zero bits in a mask.
+        /// Similar in behavior to the x86 instruction TZCNT.
+        /// </summary>
+        /// <param name="value">The value.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static uint TrailingZeroCount(int value)
+            => TrailingZeroCount(unchecked((uint)value));
+
+        /// <summary>
+        /// Count the number of trailing zero bits in a mask.
+        /// Similar in behavior to the x86 instruction TZCNT.
+        /// </summary>
+        /// <param name="value">The value.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static uint TrailingZeroCount(uint value)
+        {
+            if (Bmi1.IsSupported)
+            {
+                return Bmi1.TrailingZeroCount(value);
+            }
+
+            // TODO: See if we can leverage Log2
+
+            // Software fallback
+            // https://graphics.stanford.edu/~seander/bithacks.html#ZerosOnRightMultLookup
+            ref byte tz = ref MemoryMarshal.GetReference(TrailingCountMultiplyDeBruijn);
+            long val = (value & -value) * 0x077C_B531u;
+            uint offset = ((uint)val) >> 27;
+
+            // uint.MaxValue >> 27 is always in range [0 - 31] so we use Unsafe.AddByteOffset to avoid bounds check
+            return Unsafe.AddByteOffset(ref tz, offset);
+        }
```

Comment:
@tannergooding, I _think_ there is a bug in the original PR. 
[TZCNT](https://www.felixcloutier.com/x86/tzcnt) provides operand size as output when source operand is zero, ie 0->32.
It looks like the software fallback code here will return 0->0.

So unless the C# intrinsic is doing something different, the hardware path is different to the software fallback.
I have the (branchless) code to fix it, but wanted to confirm intent first.
(And maybe the callsites behind the original PR don't hit this border case)

Reply:
It's probably just a bug here. The `BitOps` class, whatever it does, should be consistent.

I imagine that returning `32` for `(uint)(0)` is the appropriate behavior, but it might be good to tag whoever wrote the original code.

Reply:
cc @benaadams

Reply:
Logged https://github.com/dotnet/coreclr/issues/22326

Reply:
Where its used it can never be `0`